{
    "cve_id": "CVE-2021-46880",
    "cve_description": "x509/x509_verify.c in LibreSSL before 3.4.2, and OpenBSD before 7.0 errata 006, allows authentication bypass because an error for an unverified certificate chain is sometimes discarded.",
    "cve_publish_date": "2023-04-15",
    "cwe_id": "CWE-295",
    "cwe_name": "Improper Certificate Validation",
    "cwe_description": "The product does not validate, or incorrectly validates, a certificate.",
    "commit_message": "In some situations, the verifier would discard the error on an unvalidated\ncertificte chain. This would happen when the verification callback was\nin use, instructing the verifier to continue unconditionally. This could\nlead to incorrect decisions being made in software.",
    "type_of_change": "Modification",
    "filename_of_changes": "x509_verify.c",
    "code_language": "C",
    "number_of_lines_added_for_mitigation": "83",
    "number_of_lines_deleted_vulnerable_to_cve": "46",
    "vulnerable_lines": [
        "// Line_Reference 1: /* $OpenBSD: x509_verify.c,v 1.53 2021/11/14 08:21:47 jsing Exp $ */",
        "// Line_Reference 1167: \t * Bring back the failure case we wanted to the xsc if",
        "// Line_Reference 1168: \t * we saved one.",
        "// Line_Reference 1170: \tif (!x509_verify_ctx_restore_xsc_error(ctx))",
        "// Line_Reference 1171: \t\tgoto err;",
        "// Line_Reference 1174: \t * Safety net:",
        "// Line_Reference 1175: \t * We could not find a validated chain, and for some reason do not",
        "// Line_Reference 1176: \t * have an error set.",
        "// Line_Reference 1178: \tif (ctx->chains_count == 0 && ctx->error == X509_V_OK) {",
        "// Line_Reference 1179: \t\tctx->error = X509_V_ERR_UNSPECIFIED;",
        "// Line_Reference 1180: \t\tif (ctx->xsc != NULL && ctx->xsc->error != X509_V_OK)",
        "// Line_Reference 1181: \t\t\tctx->error = ctx->xsc->error;",
        "// Line_Reference 1182: \t}",
        "// Line_Reference 1184: \t/* Clear whatever errors happened if we have any validated chain */",
        "// Line_Reference 1185: \tif (ctx->chains_count > 0)",
        "// Line_Reference 1186: \t\tctx->error = X509_V_OK;",
        "// Line_Reference 1188: \tif (ctx->xsc != NULL) {",
        "// Line_Reference 1189: \t\tctx->xsc->error = ctx->error;",
        "// Line_Reference 1190: \t\tif (ctx->chains_count > 0) {",
        "// Line_Reference 1191: \t\t\t/* Take the first chain we found. */",
        "// Line_Reference 1192: \t\t\tif (!x509_verify_ctx_set_xsc_chain(ctx, ctx->chains[0],",
        "// Line_Reference 1193: \t\t\t    1, 1))",
        "// Line_Reference 1194: \t\t\t\tgoto err;",
        "// Line_Reference 1195: \t\t\tctx->xsc->error = X509_V_OK;",
        "// Line_Reference 1196: \t\t\t/*",
        "// Line_Reference 1197: \t\t\t * Call the callback indicating success up our already",
        "// Line_Reference 1198: \t\t\t * verified chain. The callback could still tell us to",
        "// Line_Reference 1199: \t\t\t * fail.",
        "// Line_Reference 1200: \t\t\t */",
        "// Line_Reference 1201: \t\t\tif(!x509_vfy_callback_indicate_success(ctx->xsc)) {",
        "// Line_Reference 1202: \t\t\t\t/* The callback can change the error code */",
        "// Line_Reference 1203: \t\t\t\tctx->error = ctx->xsc->error;",
        "// Line_Reference 1204: \t\t\t\tgoto err;",
        "// Line_Reference 1205: \t\t\t}",
        "// Line_Reference 1206: \t\t} else {",
        "// Line_Reference 1207: \t\t\t/*",
        "// Line_Reference 1208: \t\t\t * We had a failure, indicate the failure, but",
        "// Line_Reference 1209: \t\t\t * allow the callback to override at depth 0",
        "// Line_Reference 1210: \t\t\t */",
        "// Line_Reference 1211: \t\t\tif (ctx->xsc->verify_cb(0, ctx->xsc)) {",
        "// Line_Reference 1212: \t\t\t\tctx->xsc->error = X509_V_OK;",
        "// Line_Reference 1213: \t\t\t\treturn 1;",
        "// Line_Reference 1214: \t\t\t}",
        "// Line_Reference 1217: \treturn (ctx->chains_count);",
        "// Line_Reference 1222: \tif (ctx->xsc != NULL)",
        "// Line_Reference 1223: \t\tctx->xsc->error = ctx->error;"
    ]
}
